diff --git a/tig.c b/tig.c
index 5280da8..a3d2232 100644
--- a/tig.c
+++ b/tig.c
@@ -731,7 +731,7 @@ static struct keybinding default_keybindings[] = {
 	{ KEY_UP,	REQ_PREVIOUS },
 	{ KEY_DOWN,	REQ_NEXT },
 	{ 'R',		REQ_REFRESH },
-	{ 'M',		REQ_MAXIMIZE },
+	{ 'O',		REQ_MAXIMIZE },
 
 	/* Cursor navigation */
 	{ 'k',		REQ_MOVE_UP },
@@ -3977,7 +3977,7 @@ status_open(struct view *view)
 			return FALSE;
 	}
 
-	system("git update-index -q --refresh 2>/dev/null");
+	system("git update-index -q --refresh >/dev/null 2>/dev/null");
 
 	if (!status_run(view, indexcmd, indexstatus, LINE_STAT_STAGED) ||
 	    !status_run(view, STATUS_DIFF_FILES_CMD, 0, LINE_STAT_UNSTAGED) ||
